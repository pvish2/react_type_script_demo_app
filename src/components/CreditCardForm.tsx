import * as React from "react";
import CardConstant from "../constants/Card";

interface CreditCardFormProps {
  refreshCallback: Function;
}

interface CreditCardFormState {
  name: string;
  cardNumber: number;
  limit: number;
  submitSuccess: boolean;
  errorVisible: boolean;
  message: string;
}

export default class CreditCardForm extends React.Component<
  CreditCardFormProps,
  CreditCardFormState
> {
  static defaultProps = {
    name: "string",
    cardNumber: "number",
    limit: 0,
    refreshCallback: () => {}
  };

  state = {
    name: "",
    cardNumber: 0,
    limit: 0,
    errorVisible: false,
    message: "",
    submitSuccess: false
  };

  checkValidation() {
    const { cardNumber, name, limit } = this.state;
    console.log(cardNumber, "  " + limit, "   " + name);
    if (name.length === 0) {
      this.setState({ errorVisible: true, message: CardConstant.INVALID_NAME });
      return false;
    }
    if (cardNumber.toString().length !== 10) {
      this.setState({ errorVisible: true, message: CardConstant.INVALID_CARD });

      return false;
    }
    if (limit < 0) {
      this.setState({
        errorVisible: true,
        message: CardConstant.INVALID_LIMIT
      });
      return false;
    }
    return true;
  }

  private handleSubmit = async (
    e: React.FormEvent<HTMLFormElement>
  ): Promise<void> => {
    e.preventDefault();
    if (this.checkValidation()) {
      console.log("validation success");
      this.setState({
        errorVisible: false
      });
      const submitSuccess: boolean = await this.submitForm();
      this.setState({ submitSuccess });
    }
  };

  private async submitForm(): Promise<boolean> {
    const { cardNumber, name, limit } = this.state;
    const data = {
      name: name,
      cardNumber: cardNumber.toString(),
      limit: limit
    };
    fetch(CardConstant.API_URL, {
      headers: new Headers({ "content-type": "application/json" }),
      method: "POST",
      body: JSON.stringify(data)
    })
      .then(res => res.json())
      .then(data => {
        if (data.status === 1) {
          this.setState({
            errorVisible: false,
            message: data.message,
            limit: 0
          });
          this.props.refreshCallback();
        } else {
          this.setState({ errorVisible: true, message: data.message });
        }
      })
      .catch(e => console.log(e));
    return true;
  }

  render() {
    const { message, errorVisible } = this.state;

    return (
      <div>
        <form id="form1" onSubmit={this.handleSubmit} noValidate={true}>
          <div className="inputDiv">
            <label className="label">Name</label>
            <br />
            <input
              type="text"
              className="input"
              id="name"
              required
              onChange={(e: React.FormEvent<HTMLInputElement>) =>
                this.setState({
                  name: (e.target as HTMLInputElement).value
                })
              }
            />
          </div>
          <div className="inputDiv">
            <label className="label">Card Number</label>
            <br />
            <input
              type="number"
              className="input"
              id="cardNumber"
              required
              onChange={(e: React.FormEvent<HTMLInputElement>) =>
                this.setState({
                  cardNumber: +(e.target as HTMLInputElement).value
                })
              }
            />
          </div>
          <div className="inputDiv">
            <label className="label">Limit</label>
            <br />
            <input
              type="number"
              className="input"
              id="limit"
              min="0"
              required
              onChange={(e: React.FormEvent<HTMLInputElement>) =>
                this.setState({
                  limit: +(e.target as HTMLInputElement).value
                })
              }
            />
          </div>
          <button type="submit" className="button">
            Add
          </button>
        </form>
        {errorVisible ? <h4 className="errorDiv"> {message}</h4> : null}
      </div>
    );
  }
}
