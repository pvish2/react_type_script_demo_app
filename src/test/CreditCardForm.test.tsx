import React from "react";
import ReactDOM from "react-dom";
import CreditCardForm from "../components/CreditCardForm";
import { shallow } from "enzyme";
import Constants from "../constants/Card";

describe("Credit Card Form", () => {
  let component = "";

  beforeAll(() => {
    component = shallow(<CreditCardForm />);
  });
  it("should render form", () => {
    expect(component.getElements()).toMatchSnapshot();
  });

  it("should show error on invalid card", () => {
    component.setState({ name: "bar", cardNumber: "765757577", limit: 12 });
    component.instance().checkValidation();
    expect(component.state().errorVisible).toEqual(true);
    expect(component.state().message).toEqual(Constants.INVALID_CARD);
  });

  it("should show error on invalid limit", () => {
    component.setState({ name: "bar", cardNumber: "0437093859", limit: -12 });

    component.instance().checkValidation();
    expect(component.state().errorVisible).toEqual(true);
    expect(component.state().message).toEqual(Constants.INVALID_LIMIT);
  });
});
