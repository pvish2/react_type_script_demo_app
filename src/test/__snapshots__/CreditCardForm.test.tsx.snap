// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Credit Card Form should render form 1`] = `
Array [
  <div>
    <form
      id="form1"
      noValidate={true}
      onSubmit={[Function]}
    >
      <div
        className="inputDiv"
      >
        <label
          className="label"
        >
          Name
        </label>
        <br />
        <input
          className="input"
          id="name"
          onChange={[Function]}
          required={true}
          type="text"
        />
      </div>
      <div
        className="inputDiv"
      >
        <label
          className="label"
        >
          Card Number
        </label>
        <br />
        <input
          className="input"
          id="cardNumber"
          onChange={[Function]}
          required={true}
          type="number"
        />
      </div>
      <div
        className="inputDiv"
      >
        <label
          className="label"
        >
          Limit
        </label>
        <br />
        <input
          className="input"
          id="limit"
          min="0"
          onChange={[Function]}
          required={true}
          type="number"
        />
      </div>
      <button
        className="button"
        type="submit"
      >
        Add
      </button>
    </form>
  </div>,
]
`;
